torch/csrc/autograd/python_variable_numpy.cpp

```diff
@@ -0,0 +1,107 @@
+#include "python_variable_numpy.h"
+
+#ifndef WITH_NUMPY
+namespace torch { namespace autograd {
+PyObject * THPVariable_numpy(PyObject* pyself, PyObject* arg) {
+  return PyErr_Format(PyExc_RuntimeError, "PyTorch was compiled without NumPy support");
+}
+}}
+#else
+
+#include "torch/csrc/Exceptions.h"
+#include "torch/csrc/autograd/python_variable.h"
+
+#include <ATen/ATen.h>
+#include <memory>
+#include <sstream>
+#include <stdexcept>
+
+#define NO_IMPORT_ARRAY
+#define PY_ARRAY_UNIQUE_SYMBOL __numpy_array_api
+#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
+#include <numpy/arrayobject.h>
+
+using namespace at;
+
+namespace torch { namespace autograd {
+
+static std::vector<npy_intp> cast_numpy(IntList x) {
+  auto nelem = x.size();
+  auto result = std::vector<npy_intp>(nelem);
+  for (size_t i = 0; i < nelem; i++) {
+    result[i] = static_cast<npy_intp>(x[i]);
+  }
+  return result;
+}
+
+static int numpy_dtype(const at::Type& type);
+
+PyObject * THPVariable_numpy(PyObject* pyself, PyObject* arg) {
+  HANDLE_TH_ERRORS
+  auto& self = reinterpret_cast<THPVariable*>(pyself)->cdata;
+  auto dtype = numpy_dtype(self.type());
+  auto sizes = cast_numpy(self.sizes());
+  auto strides = cast_numpy(self.strides());
+  // NumPy strides use bytes. Torch strides use element counts.
+  auto element_size_in_bytes = self.type().elementSizeInBytes();
+  for (auto& stride : strides) {
+    stride *= element_size_in_bytes;
+  }
+
+  auto array = THPObjectPtr(PyArray_New(
+      &PyArray_Type,
+      self.dim(),
+      sizes.data(),
+      dtype,
+      strides.data(),
+      self.data_ptr(),
+      0,
+      NPY_ARRAY_ALIGNED | NPY_ARRAY_WRITEABLE | NPY_ARRAY_C_CONTIGUOUS,
```

Comment:
Why `C_CONTIGUOUS`? You're not guaranteed that in here

Reply:
I had thought you wrote this, but it turns out Soumith added it in #164. He wrote, "NPY_ARRAY_C_CONTIGUOUS is not about contiguous, it's whether we are row-major or column-major".

That doesn't make any sense to me. I've looked through the NumPy [source code](https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/ctors.c), and I think that bit is ignored when you pass in strides. (The ndarray.flags is set appropriately based on the strides).